### PR TITLE
[GRPO] Sequence-level TIS & MIS

### PR DESCRIPTION
**Pull request by LeonEricsson**
# What does this PR do?
Extends the current training-inference mismatch solution from only token-level truncated importance sampling into a configurable mode allowing the user to select between:

- Token-level TIS
- Sequence-level TIS
- Token-level MIS
- Sequence-level MIS

The default is currently set to sequence-level MIS, as recent work indicates this is the sound approach, however this differs from the previous default, which may not be ideal for backward capability so I am happy to discuss that. Additionally the importance sampling _mode_ is configured via a string Literal, would appreciate thoughts on that. I was debating splitting this up into two parameters for better clarity but went with the more compact approach.

More info in #4493 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1733,7 +1753,7 @@ def _generate_and_score_completions(
         if old_per_token_logps is not None:
             output["old_per_token_logps"] = old_per_token_logps
         if self.use_vllm and self.vllm_importance_sampling_correction:
-            output["importance_sampling_ratio"] = importance_sampling_ratio
+            output["vllm_importance_sampling_ratio"] = vllm_importance_sampling_ratio
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: I’d prefer not to rename this. Doing so could break implementations that subclass this trainer. Backward compatibility isn’t mandatory, but we should let's avoid unnecessary unnecessary breaking changes.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1702,7 +1722,7 @@ def _generate_and_score_completions(
                 self.accelerator.gather(max_delta).max().item()
             )
 
-            flat_is_ratio = importance_sampling_ratio[completion_mask.bool()]
+            flat_is_ratio = token_logps_diff[completion_mask.bool()]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: `token_logps_diff` refers to `log(p_old) - log(pi_sampling)`, while `importance_sampling_ratio` refers to `p_old / pi_sampling`. So we are changing what's logged here?

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1577,10 +1578,29 @@ def _generate_and_score_completions(
...
+                    vllm_importance_sampling_ratio = torch.clamp(
+                        vllm_importance_sampling_ratio, max=self.vllm_importance_sampling_cap
+                    )
+                elif self.vllm_importance_sampling_mode in ["sequence_mask", "token_mask"]:
+                    vllm_importance_sampling_ratio = vllm_importance_sampling_ratio.masked_fill(
+                        vllm_importance_sampling_ratio > self.vllm_importance_sampling_cap, 0.0
+                    )
+                else:
+                    raise ValueError(
+                        f"Unknown vllm importance sampling level: {self.vllm_importance_sampling_mode}. Possible values are 'token_truncate', 'token_mask', 'sequence_truncate', and 'sequence_mask'."
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
                        f"Unknown vLLM importance sampling level: {self.vllm_importance_sampling_mode}. Possible values are 'token_truncate', 'token_mask', 'sequence_truncate', and 'sequence_mask'."
```

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1577,10 +1578,29 @@ def _generate_and_score_completions(
...
+                # From here, vllm_importance_sampling_ratio's shape depends on
+                # vllm_importance_sampling_mode: "token_*" level: (B, T); "sequence_*" level: (B, 1)
+
+                if self.vllm_importance_sampling_mode in ["sequence_truncate", "token_truncate"]:
+                    vllm_importance_sampling_ratio = torch.clamp(
+                        vllm_importance_sampling_ratio, max=self.vllm_importance_sampling_cap
+                    )
+                elif self.vllm_importance_sampling_mode in ["sequence_mask", "token_mask"]:
+                    vllm_importance_sampling_ratio = vllm_importance_sampling_ratio.masked_fill(
+                        vllm_importance_sampling_ratio > self.vllm_importance_sampling_cap, 0.0
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
                        vllm_importance_sampling_ratio > self.vllm_importance_sampling_cap, value=0.0
```

File: trl/trainer/grpo_config.py [hunk 5]
```diff
@@ -241,14 +241,40 @@ class GRPOConfig(TrainingArguments):
...
+                    Token-level truncated IS (default). Per-token ratios are clipped from above at C.
+                - `"token_mask"`:
+                    Token-level masked IS. Per-token ratios above C are set to zero.
+                - `"sequence_truncate"`:
+                    Sequence-level truncated IS. A single sequence ratio is clipped from above at C and
+                    applied to all tokens in the sequence.
+                - `"sequence_mask"`:
+                    Sequence-level masked IS. Sequences with ratios above C are masked out.
+
         vllm_importance_sampling_cap (`float`, *optional*, defaults to `2.0`):
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]: Is `C=2` a good default in all cases?

File: trl/trainer/grpo_config.py [hunk 6]
```diff
@@ -241,14 +241,40 @@ class GRPOConfig(TrainingArguments):
...
+                - `"token_truncate"`:
+                    Token-level truncated IS (default). Per-token ratios are clipped from above at C.
+                - `"token_mask"`:
+                    Token-level masked IS. Per-token ratios above C are set to zero.
+                - `"sequence_truncate"`:
+                    Sequence-level truncated IS. A single sequence ratio is clipped from above at C and
+                    applied to all tokens in the sequence.
+                - `"sequence_mask"`:
+                    Sequence-level masked IS. Sequences with ratios above C are masked out.
+
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]: This is a bit long for having it in the doctoring in my opinion. I'd suggest this instead:

```suggestion
        vllm_importance_sampling_mode (`str`, *optional*, defaults to `"sequence_mask"`):
            Specifies how Importance Sampling is performed when `vllm_importance_sampling_correction=True`. Possible
            values are:

                - `"token_truncate"`: Token-level truncated IS (default). Per-token ratios are clipped from above at C.
                - `"token_mask"`: Token-level masked IS. Per-token ratios above C are set to zero.
                - `"sequence_truncate"`: Sequence-level truncated IS. A single sequence ratio is clipped from above at
                  C and applied to all tokens in the sequence.
                - `"sequence_mask"`: Sequence-level masked IS. Sequences with ratios above C are masked out.
```

and create a dedicated subsection, like "Dealing with the Training-Inference Mismatch" in grpo_trainer.md, where you'd copy the above description

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: makes sense. changed the key name

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: Error on my side. The code now correctly logs the min/mean/max of the sequence- and token-level ratios. For sequence-level ratios, [we already applied masking](https://github.com/huggingface/trl/blob/19672cb3881be2c560f44fde336f8932a7ba98ad/trl/trainer/grpo_trainer.py#L1581), so we simply take all remaining sequence ratios into account. For the per-token setting, we filter out masked tokens before computing the statistics.

**Comment by LeonEricsson** [comment]
@qgallouedec have a look at the docs in grpo_trainer.md and paper_index.md, and let me know what you think.

File: docs/source/paper_index.md [hunk 7]
```diff
@@ -197,10 +197,87 @@ training_args = GRPOConfig(
...
     vllm_importance_sampling_correction=True, # default True
+    vllm_importance_sampling_mode="sequence_truncate", # or "token_truncate"
     vllm_importance_sampling_cap=2.0, # hyper-parameter C
 )
 ```
 
+### Masked Importance Sampling
+
+**📰 Blog**: https://ringtech.notion.site/icepop
+**📰 Blog**: https://yingru.notion.site/When-Speed-Kills-Stability-Demystifying-RL-Collapse-from-the-Training-Inference-Mismatch-271211a558b7808d8b12d403fd15edda
```
Comment (qgallouedec) [inline docs/source/paper_index.md:247]:
```suggestion

**📰 Blog**: https://yingru.notion.site/When-Speed-Kills-Stability-Demystifying-RL-Collapse-from-the-Training-Inference-Mismatch-271211a558b7808d8b12d403fd15edda
```

Otherwise they render on the same line

File: docs/source/paper_index.md [hunk 8]
```diff
@@ -197,10 +197,87 @@ training_args = GRPOConfig(
...
+\Bigg[
+  R(x,y)\,\cdot\,
+  \sum_{t=0}^{|y|-1}
+    \frac{\pi^{\text{training}}_\theta(y_t\,|\,x, y_{<t})}
+         {\pi^{\text{inference}}_\theta(y_t\,|\,x, y_{<t})}
+    \,\nabla_\theta \log \pi^{\text{training}}_\theta(y_t\,|\,x, y_{<t})
+\Bigg]
+$$
+
+The correct, unbiased policy gradient estimator applies a single importance ratio over the entire generated sequence (trajectory) \\( y \\), The Sequence-Level IS estimator looks like:
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
The correct, unbiased policy gradient estimator applies a single importance ratio over the entire generated sequence (trajectory)  \\( y \\), The Sequence-Level IS estimator looks like:
```

File: trl/trainer/grpo_config.py [hunk 9]
```diff
@@ -665,18 +673,38 @@ class GRPOConfig(TrainingArguments):
     vllm_importance_sampling_correction: bool = field(
         default=True,
         metadata={
-            "help": "Whether to apply Truncated Importance Sampling (TIS) between vLLM completion logprobs and "
-            "recomputed logprobs. Your Efficient RL Framework Secretly Brings You Off-Policy RL "
-            "Training highlights that using a separate generation framework (such as vLLM) can introduce off-policy "
-            "effects due to subtle implementation differences between generation and training backends. TIS is "
-            "proposed as a remedy for this issue."
+            "help": (
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]: For style consistency can you use this

```python
"help": "..."
"..."
```

instead of 

```python
"help": (
    "..."
    "..."
)
```

**Review by qgallouedec — APPROVED** [review]
Perfect! Very nice and clean addition. I only have minor stylistic remarks.

File: trl/trainer/grpo_config.py (hunk 5 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_config.py]: The original TIS blog is token-level and uses `C=2`. IcePop (from Ling Team) is also token-level and uses`C=5`. Qwen Team's blog uses `C=2` for all experiments covering MIS/TIS and sequence/token. The FP16 paper uses `C=3`. 

I would set the default higher than 2. A more permissive upper bound interferes less, so it is safer to start high and let users tune it down. Somewhere in the 3-5 range.

Comment (qgallouedec) [reply trl/trainer/grpo_config.py]: I don't have strong opinion on this. Feel free to modify the default if you think it makes sense

**Comment by qgallouedec** [comment]
test are passing, feel free to merge when you feel it's ready

**Comment by lzy37ld** [comment]
Any plan to merge?

**Comment by LeonEricsson** [comment]
> Any plan to merge?

I was hoping to run some validation training runs before merging, but haven't had the spare time to do so.
